facebook/ThreatExchange
# Update experimentalfetch to work with the new version of threat_updates

Summary
---------

Update experimentalfetch to work with the new version of threat_updates. Make experimentalfetch more resilient by adding error handling and more thorough checkpointing. Update the indicator "storage system" to no longer rely on signal types and instead store based on indicator type.

Test Plan
---------

1. verify fetch still works:
Ran: "threatexchange -c ~/large_dataset_collab.te.cfg fetch --until-timestamp `date -j -vMarm -v1d -v2018y '+%s'`" to competition.
Also used fetch to query the entire large dataset set.

2. Verify experimentalfetch gets everything:
Ran: "threatexchange -c ~/large_dataset_collab.te.cfg experimental-fetch" on the large dataset collaboration and it queried to completion getting ~1.93m records

3. Verify experimentalfetch  --start-time and --end-time
Ran with various combinations of --start-time and --end-time and things worked as expected (only data updated within the time range was returned)

4. Verify experimentalfetch --threat-types
Ran "threatexchange -c ~/other_collab.te.cfg experimental-fetch --threat-types HASH_PDQ URL" on another collaboration and things worked as expected (only data with indicator types of HASH_PDQ or URL were returned)

5. Verify experimentalfetch --continuation
Ran "threatexchange -c ~/other_collab.te.cfg experimental-fetch --continuation" and this works as described:
i.e.
* When no url/cursor is check-pointed threat_updates is queried using the check-pointed last_stop_time as the new start-time and the check-pointed indicator types (if given) as filters
* When a url/cursor is check-pointed we continue fetching from that url/cursor

6. Verify experimentalfetch forces a full update if you haven't fetched in 85 days
* Modified the checkpoint to be > 85 days old
* Tried to run with --continuation
* Got the following message "It's been a long time since a full fetch, forcing one now." and started to fetch the entire collaboration

7. Tested error handling by both explicitly throwing errors and mucking with the url being queried
